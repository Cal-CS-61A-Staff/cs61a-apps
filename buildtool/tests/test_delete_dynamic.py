from harness import AddDep, Environment, Input, Sh, create_test_env


def test_deleting_dynamic_deps(snapshot):
    with create_test_env(snapshot) as env:
        env: Environment

        env.annotate("Create two files: f1 and f2")
        env.annotate("Initially, f1 will dynamically depend on f2")
        f1 = env.new_file()
        f2 = env.new_file()
        rule = env.declare_rule(
            AddDep(f1),
            input_action := Input([f1, f2], [AddDep(f2)]),
            shell_action := Sh([f1, f2]),
        )
        env.build(rule)

        env.annotate(
            "We then modify f1 so it no longer depends on f2, and also delete f2 altogether"
        )
        env.update_file(f1)
        env.delete_file(f2)
        input_action.update_result([f1], [])
        shell_action.update_result([f1])

        env.annotate(
            "Now, when we build the rule, the cached input will try to look up f2, but will fail to do so."
        )
        env.annotate(
            "However, this error should be caught and ignored, because dynamic deps are not guaranteed to be correct"
        )
        env.build(rule)
