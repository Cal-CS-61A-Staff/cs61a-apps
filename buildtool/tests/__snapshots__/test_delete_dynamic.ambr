# name: test_deleting_dynamic_deps
  '
  STARTING TEST
  
  # Create two files: f1 and f2
  # Initially, f1 will dynamically depend on f2
  CREATING FILE f1.txt with contents: "V0 (f1.txt)"
  CREATING FILE f2.txt with contents: "V0 (f2.txt)"
  CREATING rule RuleA
  
  BUILDING rule RuleA
  RUNNING run.py --mode=INPUT --id=0 in working directory
  RUNNING run.py --mode=INPUT --id=0 in scratch directory with files ['f1.txt', 'f2.txt']
  RUNNING run.py --mode=SH --id=1 in scratch directory with files ['f1.txt', 'f2.txt']
  COMPLETED BUILDING rule RuleA
  
  # We then modify f1 so it no longer depends on f2, and also delete f2 altogether
  UPDATING FILE f1.txt with contents: "V1 (f1.txt)"
  DELETING FILE f2.txt
  
  # Now, when we build the rule, the cached input will try to look up f2, but will fail to do so.
  # However, this error should be caught and ignored, because dynamic deps are not guaranteed to be correct
  BUILDING rule RuleA
  RUNNING run.py --mode=INPUT --id=0 in working directory
  RUNNING run.py --mode=INPUT --id=0 in scratch directory with files ['f1.txt']
  RUNNING run.py --mode=SH --id=1 in scratch directory with files ['f1.txt']
  COMPLETED BUILDING rule RuleA
  
  
  '
---
